File: C++/Trees/Binary Search Trees/traversals.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};

/*
    Depth First Search (DFS)
        - Preorder : <root> <left> <right>
        - Inorder  : <left> <root> <right> 
        - Postorder: <left> <right> <root>  

*/

void printPreorder(Node* root) {
    if(root == NULL) return;
    cout << root->data << " ";
    printPreorder(root->left);
    printPreorder(root->right); 
}

void printInorder(Node* root) {
    if(root == NULL) return;
    printInorder(root->left);
    cout << root->data << " ";
    printInorder(root->right);
}

void printPostorder(Node* root) {
    if(root == NULL) return;
    printPostorder(root->left);
    printPostorder(root->right);
    cout << root->data << " ";
}


int main() {
    struct Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
/*
      1
    /   \
   2     3
  / \   / \
  4  5 6   7
*/
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);

    printInorder(root);
}

File: Python/Binary Tree/traversals.py
from queue import Queue

class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def inOrder(root):
    if root is None:
        return
    inOrder(root.left)
    print(root.val, end=' ')
    inOrder(root.right)

def preOrder(root):
    if root is None:
        return
    print(root.val, end=' ')
    preOrder(root.left)
    preOrder(root.right)

def postOrder(root):
    if root is None:
        return
    postOrder(root.left)
    postOrder(root.right)
    print(root.val, end=' ')

def levelOrder(root):
    q = Queue()
    q.put(root)
    while not q.empty():
        temp = q.get()
        print(temp.val, end=' ')
        if temp.left:
            q.put(temp.left)
        if temp.right:
            q.put(temp.right)

# #       1
# #    2    3
# #   4       5
# root = TreeNode(1)
# root.left = TreeNode(2)
# root.right = TreeNode(3)
# root.left.left = TreeNode(4)
# root.left.right = TreeNode(5)
# print("Inorder:")
# inOrder(root)
# print()

# print("PostOrder:")
# postOrder(root)
# print()

# print("PreOrder:")
# preOrder(root)
# print()

# print("LevelOrder:")
# levelOrder(root)
# print()
